### PR TITLE
[FEATURE] Add support of link filtering to contact sensors.

### LINKED ISSUE #2771: [Feature]: Add filter_link_idx to ContactForceSensor
### What feature or enhancement are you proposing?

Add `filter_link_idx` to `ContactForceSensor`, mirroring the parameter `ContactSensor` in #2655.

`ContactSensor` can filter its boolean by counterpart link (blacklist), but `ContactForceSensor` reports the total contact force with no way to scope it.

Proposal: add `filter_link_idx: array-like[int]` to `gs.sensors.ContactForce` with the same semantics as `gs.sensors.Contact`: contacts whose counterpart link is in the list are excluded from the reported force.

### Motivation

Wanted a `ContactForceSensor` that reports only the contact force from a specific subset of counterpart links (e.g. "force on the foot from the ground, excluding self-contact" or "force on the torso from anything except the robot itself"),

### Potential Benefit

Reliable reward functions in RL

### What is the expected outcome of the implementation work?

- `gs.sensors.ContactForce` accepts `filter_link_idx: array-like[int]`, same semantics as `gs.sensors.Contact`.
- Reported force excludes contacts whose counterpart link is in the list, per-contact (correct even when the link touches multiple counterparts simultaneously).

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jsw7460**
ContactForceSensor now accepts filter_link_idx, mirroring ContactSensor: contacts with the sensor link whose other participant is one of the listed links are excluded from the reported force. The filter is applied on both ground-truth-update paths; the vectorized zerocopy path and the _kernel_get_contacts_forces kernel. 
Adds test_contact_force_sensor_filter_link_idx.

 ## Description

  `ContactForceSensor` now accepts `filter_link_idx` (default `()`), mirroring
  `ContactSensor` from #2655: contacts with the sensor link whose **other**
  participant is one of the listed links are excluded from the reported force.

    - `_kernel_get_contacts_forces` (the kernel fallback): takes
      `filter_links_idx` and skips a contact for a sensor when the counterpart
      is in that sensor's filter list;
    - `filtered_sensor_idx` gates the comparison so sensors without a filter
      keep the cheap path (same optimisation as `ContactSensor`).

  ## Related Issue

  Resolves Genesis-Embodied-AI/Genesis#2771

  ## Motivation and Context

  `ContactSensor` already filters its boolean by counterpart link, but `ContactForceSensor` only reported the total contact force on the link, with no way to scope it (e.g. "force on a foot from the ground, excluding self-contact"). 

  ## How Has This Been / Can This Be Tested?

      pytest tests/test_sensors.py::test_contact_force_sensor_filter_link_idx tests/test_sensors.py::test_contact_sensor_filter_link_idx

  `test_contact_force_sensor_filter_link_idx` puts a box on the floor (env 0) /  a box on the floor with another box stacked on top (env 1), with two `ContactForce` sensors on the bottom box — one plain, one with `filter_link_idx (floor.link_start,)`. It asserts the filtered sensor reads ~0 in env 0 and reads only the top box's downward push in env 1,  while the unfiltered sensor reads the net contact force in both. 

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [x] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [ x] I tested my changes and added instructions on how to test it for reviewers.

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Milotrince** [comment]
This feature has already been implemented by https://github.com/Genesis-Embodied-AI/genesis-world/pull/2655 .

**Comment by jsw7460** [comment]
@Milotrince  
Hi, thank you for #2655 !
But #2655 added `filter_link_idx` only to `gs.sensors.Contact` (the boolean contact sensor), not to `gs.sensors.ContactForce` (the force-vector sensor).

What I'm after here is a `ContactForce` sensor that sums only the force from contacts with a given subset of links — e.g. "force on a foot from the ground, excluding self-contact" — which
  isn't possible today since `ContactForce` reports the total contact force on the link.

**Comment by Milotrince** [comment]
I rebased your branch and updated to extend the filtering to ContactDepthProbe and KinematicTaxel as well as ContactProbe!

**Review by Milotrince — DISMISSED** [review]

File: genesis/utils/misc.py [hunk 1]
```diff
@@ -489,6 +489,22 @@ def _default_factory():
...
+    ``-1`` is a sentinel that never matches a real link index, so kernels can scan every column unconditionally. The
+    table keeps at least one column so that even an all-empty table stays a valid (non-zero-dim) kernel argument.
+    """
+    n_sensors, cur_max = filter_links_idx.shape
+    new_max = max(cur_max, len(filter_link_idx), 1)
+    out = torch.full((n_sensors + 1, new_max), -1, dtype=gs.tc_int, device=gs.device)
+    out[:n_sensors, :cur_max] = filter_links_idx
+    if len(filter_link_idx) > 0:
+        out[n_sensors, : len(filter_link_idx)] = torch.tensor(filter_link_idx, dtype=gs.tc_int, device=gs.device)
+    return out
```
Comment (duburcqa) [inline genesis/utils/misc.py]: I don't think it is the right place for such helper. it is not generic enough.

File: genesis/options/sensors/options.py [hunk 2]
```diff
@@ -346,12 +356,7 @@ class Contact(RigidSensorOptionsMixin["ContactSensor"], SimpleSensorOptions["Con
 
     def validate_scene(self, scene: "Scene"):
         super().validate_scene(scene)
-        if self.filter_link_idx:
-            n_links = scene.sim.rigid_solver.n_links
-            if np.any(np.array(self.filter_link_idx) < 0) or np.any(np.array(self.filter_link_idx) >= n_links):
-                gs.raise_exception(
-                    f"Contact sensor filter_link_idx should be in range [0, {n_links}). Got {self.filter_link_idx}"
-                )
+        _validate_filter_link_idx(scene, self.filter_link_idx, "Contact sensor")
```
Comment (duburcqa) [inline genesis/options/sensors/options.py]: Remove '"Contact sensor"'

File: genesis/options/sensors/options.py
Comment (duburcqa) [inline genesis/options/sensors/options.py]: This has no place in generic sensor option. There is no notion on link at this stage.

File: genesis/utils/misc.py (hunk 1 above)
Comment (duburcqa) [reply genesis/utils/misc.py]: Moved off misc.py into `ContactFilterMetadataMixin.append_filter` in contact_force.py, next to the sensors that own the filter table.

File: genesis/options/sensors/options.py (hunk 2 above)
Comment (duburcqa) [reply genesis/options/sensors/options.py]: Done -- the range-check message now derives the sensor name from `type(self).__name__` instead of a hard-coded label.

File: genesis/options/sensors/options.py
Comment (duburcqa) [reply genesis/options/sensors/options.py]: Moved into a dedicated `ContactFilterOptionsMixin` (a rigid-contact-sensor mixin) shared by the contact/force/tactile sensors; the generic sensor options no longer reference links.

**Comment by duburcqa** [comment]
@codex review

File: genesis/engine/sensors/kinematic_tactile.py [hunk 3]
```diff
@@ -515,9 +523,11 @@ def _kernel_build_sensor_candidate_geom_mask(
         n_c = sensor_n_contacts[i_b, i_s]
         for i_c_ in range(n_c):
             i_c = sensor_contacts_idx[i_b, i_s, i_c_]
-            if collider_state.contact_data.link_a[i_c, i_b] == link:
+            c_link_a = collider_state.contact_data.link_a[i_c, i_b]
+            c_link_b = collider_state.contact_data.link_b[i_c, i_b]
+            if c_link_a == link and not _func_link_is_filtered(filter_links_idx, i_s, c_link_b):
                 sensor_candidate_geom_mask[i_b, i_s, collider_state.contact_data.geom_b[i_c, i_b]] = True
-            if collider_state.contact_data.link_b[i_c, i_b] == link:
+            if c_link_b == link and not _func_link_is_filtered(filter_links_idx, i_s, c_link_a):
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/sensors/kinematic_tactile.py]: fixed

**Review by duburcqa — DISMISSED** [review]

**Comment by duburcqa** [comment]
@codex review

**Review by duburcqa — APPROVED** [review]
